component/gnap/as/client.go

```diff
@@ -0,0 +1,203 @@
+/*
+Copyright SecureKey Technologies Inc. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package as
+
+import (
+	"bytes"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+
+	"github.com/trustbloc/edge-core/pkg/log"
+
+	gnaprest "github.com/trustbloc/auth/pkg/restapi/gnap"
+	"github.com/trustbloc/auth/spi/gnap"
+)
+
+// package as is the Authentication Server client that requests GNAP tokens from the Authorization Server.
+
+//nolint:gochecknoglobals
+var logger = log.New("auth-server-client")
+
+const contentType = "application/json"
+
+// Client requesting Gnap tokens from the Authorization Server.
+type Client struct {
+	signer            gnap.Signer
+	httpClient        *http.Client
+	gnapAuthServerURL string
+}
+
+// NewClient creates a new GNAP authorization client. It requires a signer for HTTP Signature header, an HTTP client
+// and a base URL of the authorization server.
+func NewClient(signer gnap.Signer, httpClient *http.Client, gnapAuthServerURL string) (*Client, error) {
+	if signer == nil {
+		return nil, fmt.Errorf("gnap auth client: missing signer")
+	}
+
+	if httpClient == nil {
+		return nil, fmt.Errorf("gnap auth client: missing http client")
+	}
+
+	if gnapAuthServerURL == "" {
+		return nil, fmt.Errorf("gnap auth client: missing Authorization Server URL")
+	}
+
+	return &Client{
+		signer:            signer,
+		httpClient:        httpClient,
+		gnapAuthServerURL: gnapAuthServerURL,
+	}, nil
+}
+
+// RequestAccess creates a GNAP grant access req then submit it to the server to receive a response with an
+// interact_ref value.
+//nolint:funlen,gocyclo
+func (c *Client) RequestAccess(req *gnap.AuthRequest) (*gnap.AuthResponse, error) {
+	if req == nil {
+		return nil, fmt.Errorf("requestAccess: empty request")
+	}
+
+	mReq, err := json.Marshal(req)
+	if err != nil {
+		return nil, fmt.Errorf("requestAccess: marshal access token error: %w", err)
+	}
+
+	var (
+		sig               []byte
+		marshalledRequest []byte
+	)
+
+	sig, err = c.signer.Sign(mReq)
+	if err != nil {
+		return nil, fmt.Errorf("requestAccess: signature error: %w", err)
+	}
+
+	marshalledRequest, err = json.Marshal(req)
```

Comment:
Why do we need to marshal req twice ? I see marshal in [line 67](https://github.com/trustbloc/auth/pull/163/files#diff-b8608db5d8c65f68e3b890d198c90536f3ccb174f3816f8899af0a4d5f3f4586R67) as well.

Reply:
the signer got me distracted :)

i'll remove..

Reply:
done